File: src/main/java/org/saigon/striker/service/UserService.java
package org.saigon.striker.service;

import org.saigon.striker.model.UserEntity;
import org.saigon.striker.model.UserRepository;
import org.springframework.security.core.userdetails.ReactiveUserDetailsService;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import static org.apache.commons.lang3.Validate.notEmpty;
import static org.apache.commons.lang3.Validate.notNull;

@Service
public class UserService implements ReactiveUserDetailsService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    @Override
    public Mono<UserDetails> findByUsername(String username) {
        notEmpty(username);
        return userRepository.findByUsername(username)
                .map(this::userEntityToUserDetails);
    }

    public Mono<UserEntity> createUser(UserEntity userEntity) {
        notNull(userEntity);
        return userRepository.save(userEntity.withPassword(
                passwordEncoder.encode(userEntity.getPassword())));
    }

    public Mono<UserEntity> getUser(String id) {
        notEmpty(id);
        return userRepository.findById(id);
    }

    public Mono<UserEntity> getUserByUsername(String username) {
        notEmpty(username);
        return userRepository.findByUsername(username);
    }

    public Mono<Void> deleteUser(String id) {
        notEmpty(id);
        return userRepository.deleteById(id);
    }

    private UserDetails userEntityToUserDetails(UserEntity userEntity) {
        return org.springframework.security.core.userdetails.User.builder()
                .username(userEntity.getUsername())
                .password(userEntity.getPassword())
                .roles(UserRoles.USER.toString())
                .build();
    }
}

File: src/main/kotlin/org/saigon/striker/service/UserService.kt
package org.saigon.striker.service

import org.saigon.striker.model.UserEntity
import org.saigon.striker.model.UserRepository
import org.saigon.striker.model.UserRoles
import org.springframework.security.core.userdetails.ReactiveUserDetailsService
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
class UserService(
    val userRepository: UserRepository,
    val passwordEncoder: PasswordEncoder
) : ReactiveUserDetailsService {

    override fun findByUsername(username: String?): Mono<UserDetails> {
        return userRepository.findByUsernameMono(username ?: throw IllegalArgumentException())
            .map(UserEntity::toUserDetails)
    }

    suspend fun createUser(userEntity: UserEntity): UserEntity {
        if (userRepository.findByUsername(userEntity.username) != null) {
            throw UsernameAlreadyExistsException(userEntity.username)
        }
        return userRepository.save(userEntity.encodePassword(passwordEncoder))
    }

    suspend fun getUser(id: String): UserEntity? = userRepository.findById(id)

    suspend fun getUserByUsername(username: String): UserEntity? = userRepository.findByUsername(username)

    suspend fun deleteUser(id: String): Unit = userRepository.deleteById(id)
}

private fun UserEntity.toUserDetails(): UserDetails = User.builder()
    .username(username)
    .password(password)
    .roles(UserRoles.USER.toString())
    .build()

private fun UserEntity.encodePassword(passwordEncoder: PasswordEncoder): UserEntity =
    copy(password = passwordEncoder.encode(password))
